aiida/common/hashing.py

```diff
@@ -234,17 +234,35 @@ def _(val, **kwargs):
     return [_single_digest('none')]
 
 
-@_make_hash.register(datetime)
+@_make_hash.register(datetime.datetime)
 def _(val, **kwargs):
     """hashes the little-endian rep of the float <epoch-seconds>.<subseconds>"""
     # see also https://stackoverflow.com/a/8778548 for an excellent elaboration
     if val.tzinfo is None or val.utcoffset() is None:
         val = val.replace(tzinfo=pytz.utc)
-    timestamp = val.timestamp()
 
+    timestamp = val.timestamp()
     return [_single_digest('datetime', float_to_text(timestamp, sig=AIIDA_FLOAT_PRECISION).encode('utf-8'))]
 
 
+@_make_hash.register(datetime.date)
+def _(val, **kwargs):
+    """hashes the little-endian rep of the float <epoch-seconds>.<subseconds>"""
+    import calendar
+    from aiida.manage.configuration import settings
+
+    # The method to get the timestamp from the date object depends on whether it is in UTC or a local timezeon
+    # See for an extended discussion https://stackoverflow.com/a/8778548
+    if getattr(settings, 'USE_TZ', False):
```

Comment:
I'm not sure if TZ handling makes sense here. Without it, we could just use ``_single_digest('date', val.isoformat().encode('utf-8'))``.

In addition to the QueryBuilder, `_make_hash` is also used in the caching implementation. Since the `datetime.date` itself is a naive, timezone-unaware datastructure, I think the hashing should do the same. Both should be more or less trouble-free in most cases. But, if I were to import a `Node` containing a `datetime.date` somewhere (if that's even possible?) into a system using a different timezone, caching for that node would be broken.

Reply:
> Since the `datetime.date` itself is a naive, timezone-unaware datastructure, I think the hashing should do the same. Both should be more or less trouble-free in most cases. 

Fair enough

> But, if I were to import a `Node` containing a `datetime.date` somewhere (if that's even possible?) into a system using a different timezone, caching for that node would be broken.

When talking about `attributes` as to what a `Node` can contain, then it has to be serializable according to `aiida.orm.utils.node.clean_value` and neither `datetime.datetime` nor `datetime.date` objects are supported. The former used to be, but we removed that. Users are now required to serialize these themselves into a serializable type, such as a timestamp. The deserialization also has to be done manually by the caller. So I guess it is true, even for the caching mechanism this shouldn't matter because it should never get a `datetime.date` object but some string or float timestamp.

@giovannipizzi do you agree that date objects we can always treat as timezone naive and just take the isoformat string?

Reply:
Discussed in person with @giovannipizzi and he agrees that timezone for `date` object is meaningless and so just the ISO formatted string is fine